File: app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/adapters/db/dao/AttachmentDao.java
package it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;

import java.util.List;

import it.niedermann.nextcloud.deck.model.Attachment;

@Dao
public interface AttachmentDao extends GenericDao<Attachment> {
    @Query("SELECT * FROM attachment where cardId = :cardId")
    LiveData<List<Attachment>> getAttachmentsForCard(long cardId);

    @Query("SELECT * FROM attachment where accountId = :accountId and id = :remoteId")
    Attachment getAttachmentByRemoteIdDirectly(long accountId, Long remoteId);

    @Query("SELECT * FROM attachment where accountId = :accountId and localId = :id")
    Attachment getAttachmentByLocalIdDirectly(long accountId, Long id);

    @Query("SELECT * FROM attachment WHERE accountId = :accountId and cardId = :localCardId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<Attachment> getLocallyChangedAttachmentsByLocalCardIdDirectly(long accountId, long localCardId);

    @Query("SELECT * FROM attachment WHERE accountId = :accountId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<Attachment> getLocallyChangedAttachmentsDirectly(long accountId);

    @Query("SELECT a.* FROM attachment a inner join card c on c.localId = a.cardId " +
            "WHERE c.stackId = :localStackId and (a.status<>1 or a.id is null or a.lastModified <> a.lastModifiedLocal)")
    List<Attachment> getLocallyChangedAttachmentsForStackDirectly(long localStackId);

    @Query("SELECT * FROM attachment WHERE accountId = :accountId and cardId = :localCardId")
    List<Attachment> getAttachmentsForLocalCardIdDirectly(long accountId, Long localCardId);
}

File: data/local/src/main/java/it/niedermann/nextcloud/deck/data/local/dao/AttachmentDao.kt
package it.niedermann.nextcloud.deck.data.local.dao

import androidx.room3.Dao
import androidx.room3.Insert
import androidx.room3.OnConflictStrategy
import androidx.room3.Query
import io.reactivex.rxjava3.core.Flowable
import it.niedermann.nextcloud.deck.data.local.entity.AttachmentEntity
import java.util.concurrent.CompletableFuture

@Dao
interface AttachmentDao : GenericDao<AttachmentEntity> {

    @Query("SELECT * FROM Attachment WHERE cardId = :cardId")
    fun getAttachmentsByCard(cardId: Long): Flowable<List<AttachmentEntity>>

    @Query("SELECT * FROM Attachment WHERE accountId = :accountId AND status != 1")
    fun getChangedAttachments(accountId: Long): CompletableFuture<List<AttachmentEntity>>

    @Query("SELECT * FROM Attachment WHERE accountId = :accountId AND remoteId = :remoteId")
    fun getAttachmentByRemoteId(accountId: Long, remoteId: Long): CompletableFuture<AttachmentEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertOrReplace(entity: AttachmentEntity): CompletableFuture<Long>

    @Query("DELETE FROM Attachment WHERE localId = :localId")
    fun deleteById(localId: Long): CompletableFuture<Void?>
}
